api/core/rag/retrieval/dataset_retrieval.py

```diff
@@ -557,12 +562,25 @@ def multiple_retrieve(
                             "score_threshold": score_threshold,
                             "query": None,
                             "attachment_id": attachment_id,
+                            "cancel_event": cancel_event,
+                            "thread_exceptions": thread_exceptions,
                         },
                     )
                     all_threads.append(attachment_thread)
                     attachment_thread.start()
-            for thread in all_threads:
-                thread.join()
+
+            # Poll threads with short timeout to detect errors quickly (fail-fast)
+            while any(t.is_alive() for t in all_threads):
+                for thread in all_threads:
+                    thread.join(timeout=0.1)
+                    if thread_exceptions:
+                        cancel_event.set()
+                        break
+                if thread_exceptions:
+                    break
+
+            if thread_exceptions:
+                raise thread_exceptions[0]
```

Comment:
Maybe we can utilize https://docs.python.org/3/library/exceptions.html#ExceptionGroup to keep all the exception information?

Example:

```py
try:
    raise ValueError("oops")
except Exception as exc:
    ex1 = exc


try:
    raise TypeError("oops")
except Exception as exc:
    ex2 = exc


raise ExceptionGroup("Oops", [ex1, ex2])
```

<img width="967" height="349" alt="Image" src="https://github.com/user-attachments/assets/abbf0d3e-ddf7-4f59-aafe-092785ca0539" />

There is [a tutorial for `ExceptionGroup`](https://docs.python.org/3/tutorial/errors.html#tut-exception-groups) in Python documentation.